crates/cxx-qt-build/src/lib.rs

```diff
@@ -536,110 +555,129 @@ impl CxxQtBuilder {
         // Write cxx-qt and cxx headers
-        cxx_qt::write_headers(format!("{header_root}/cxx-qt"));
-        std::fs::create_dir_all(format!("{header_root}/rust"))
+        cxx_qt::write_headers(header_root.join("cxx-qt"));
+        std::fs::create_dir_all(header_root.join("rust"))
             .expect("Could not create cxx header directory");
-        let h_path = format!("{header_root}/rust/cxx.h");
+        let h_path = header_root.join("rust").join("cxx.h");
         // Wrap the File in a block scope so the file is closed before the compiler is run.
         // Otherwise MSVC fails to open cxx.h because the process for this build script already has it open.
         {
-            let mut header = File::create(h_path).expect("Could not create cxx.h");
-            write!(header, "{}", cxx_gen::HEADER).expect("Could not write cxx.h");
+            std::fs::write(h_path, cxx_gen::HEADER).expect("Failed to write cxx.h");
         }
+    }
 
-        // Setup compiler
-        // Static QML plugin and Qt resource initialization need to be linked with +whole-archive
-        // because they use static variables which need to be initialized before main
-        // (regardless of whether main is in Rust or C++). Normally linkers only copy symbols referenced
-        // from within main when static linking, which would result in discarding those static variables.
-        // Use a separate cc::Build for the little amount of code that needs to be linked with +whole-archive
-        // to avoid bloating the binary.
-        let mut cc_builder_whole_archive = cc::Build::new();
-        cc_builder_whole_archive.link_lib_modifier("+whole-archive");
-
-        // Ensure we are not using rustc 1.69
-        if let Some(version) = version_check::Version::read() {
-            let (major, minor, _) = version.to_mmp();
-            if major == 1 && minor == 69 {
-                // rustc 1.69 had a regression where +whole-archive wouldn't
-                // work without specifying -bundle.
-                // https://github.com/rust-lang/rust/pull/110917
-                //
-                // However, we need to not have -bundle for qt-static-initializers to work
-                // with CMake builds, otherwise the statement below occurs where it's missing
-                // from the final binary.
-                //
-                // When building a staticlib -bundle means that the native static library
-                // is simply not included into the archive and some higher level build
-                // system will need to add it later during linking of the final binary.
-                // https://doc.rust-lang.org/rustc/command-line-arguments.html#option-l-link-lib
-                panic!("rustc 1.69.x is not supported with CXX-Qt due to a compiler bug.\nSee: https://github.com/rust-lang/rust/pull/110917\nPlease update your compiler using 'rustup update' or use an older compiler.");
-            }
+    fn setup_cc_builder<'a>(
```

Comment:
this might be useful to have public and hidden, as note that cxx-qt needs to setup things in a similar way

Reply:
Right, this could maybe go into a `cxx-qt-build-shared` crate or similar.
Currently cxx-qt-build depends on cxx-qt, so we can't yet do this unfortunately.